src/app/server/Server.cpp

```diff
@@ -502,3 +502,18 @@ FabricTable & GetGlobalFabricTable()
 {
     return gFabrics;
 }
+
+ExchangeManager & GetGlobalExchangeManager()
+{
+    return gExchangeMgr;
+}
+
+SessionIDAllocator & GetGlobalSessionIDAllocator()
+{
+    return gSessionIDAllocator;
+}
+
+SecureSessionMgr & GetGlobalSessionManager()
+{
+    return gSessions;
+}
```

Comment:
Is there a cleaner way than just having accessors for singletons? Seems like we could inject there rather than have more singleton accessors.

Reply:
Can you give a detailed description on the injection? Thanks.

Reply:
In the code snippet you have:

```
   SecureSessionMgr &sessionMgr = GetGlobalSessionManager();
        chip::Messaging::ExchangeManager &exchangeMgr = GetGlobalExchangeManager();
        chip::Controller::ControllerDeviceInitParams params = {nullptr,          &sessionMgr,
                                                               &exchangeMgr,     nullptr,
                                                               nullptr,          &mOperationalCredentialSet,
                                                               mCredentialIndex, &GetGlobalSessionIDAllocator(),
                                                               nullptr};
        mDevice.Init(params, CHIP_PORT, mPeerNodeId, mPeerAddr, mFabricIndex);
        mDevice.MarkOperationalCertProvisioned();
```

Curious why:

1. GetGlobal* must be called, rather than already having access to the object references at the call site (e.g. previously injected dependency)
2. `mDevice.MarkOperationalCertProvisioned()` is a workaround around the fact that CHIPDeviceController has too tight a coupling to the commissioning cases. Suggest refactoring the underlying issue, rather than "pretending a commisisoning was done" when the class already has too many ties to some disjoint flows

Reply:
The `GetGlobal*` class are added since these objects are located in `app/server` and it's hard to determine the device to be added there. For instance we may want to scan for a device first and then create a control session.

Reply:
I think the issue here is that Server.cpp instantiates its own singleton classes like `ExchangeManager`. Instead, we could initialize Server.cpp with an `ExchangeManager` (and other singletons) and those would then be accessible for the application to use to initialize a `Device`. Is this aligned with what you're suggesting @tcarmelveilleux ?

Reply:
Moving these singletons are possible but requires broad changes in the examples. I'll remove the `GetGlobal` changes from this PR and add a separate PR to resolve this issue.

Reply:
I believe dependency injection should be vastly preferred to global singletons. Singletons cause tight coupling and make testing a lot more difficult.

Reply:
Moved to https://github.com/project-chip/connectedhomeip/pull/9200